nameko_sqlalchemy/transaction_retry.py

```diff
@@ -1,28 +1,61 @@
+
+import eventlet
 import functools
 import operator
 
 import wrapt
 from sqlalchemy import exc
 
 
-def transaction_retry(wrapped=None, session=None):
+def transaction_retry(wrapped=None, session=None, retries=1, backoff_factor=0):
 
     if wrapped is None:
-        return functools.partial(transaction_retry, session=session)
+        return functools.partial(
+            transaction_retry, session=session,
+            retries=retries, backoff_factor=backoff_factor)
 
     @wrapt.decorator
     def wrapper(wrapped, instance, args, kwargs):
 
-        try:
-            return wrapped(*args, **kwargs)
-        except exc.OperationalError as exception:
-            if exception.connection_invalidated:
-                if isinstance(session, operator.attrgetter):
-                    session(instance).rollback()
-                elif session:
-                    session.rollback()
+        @retry(retries, backoff_factor, exc.OperationalError)
+        def run_or_rollback():
+            try:
                 return wrapped(*args, **kwargs)
-            else:
+            except exc.OperationalError as exception:
+                if exception.connection_invalidated:
+                    if isinstance(session, operator.attrgetter):
+                        session(instance).rollback()
+                    elif session:
+                        session.rollback()
                 raise
 
+        return run_or_rollback()
+
     return wrapper(wrapped)  # pylint: disable=E1120
+
+
+def retry(retries, backoff_factor, exceptions):
+
+    retries = max(retries, 1)
+    backoff_factor = max(backoff_factor, 0)
+
+    @wrapt.decorator
+    def wrapper(wrapped, instance, args, kwargs):
+        errors = 0
+        while True:
+            try:
+                return wrapped(*args, **kwargs)
+            except exceptions:
+                errors += 1
+
+                if errors > retries:
+                    raise
+
+                if errors >= 2:
+                    backoff_value = backoff_factor * (2 ** (errors - 2))
```

Comment:
First retry is always immediate. Second retry is always after `backoff_factor` seconds. Subsequent retry delays are the `backoff_factor` multiplied by an increasing power of two. Is that correct?

Reply:
yes, correct, the algorithm is a standard exponential backoff.  In essence this implementation should be no different than the algorithm nameko uses with the exception of the initial "immediate" retry (see below).